Content.Server/Explosion/EntitySystems/ExplosionSystem.GridMap.cs

```diff
@@ -0,0 +1,374 @@
+
+        foreach (var tileRef in grid.GetAllTiles())
+        {
+            if (tileRef.Tile.IsEmpty)
+                continue;
+
+            if (IsEdge(grid, tileRef.GridIndices, out var dir))
+                edges.Add(tileRef.GridIndices, dir);
+        }
+    }
+
+    private void OnGridRemoved(GridRemovalEvent ev)
+    {
+        _airtightMap.Remove(ev.GridId);
+        _gridEdges.Remove(ev.GridId);
+    }
+
+    /// <summary>
+    ///     Take our map of grid edges, where each is defHined in their own grid's reference frame, and map those
+    ///     edges all onto one grids reference frame.
+    /// </summary>
+    public (Dictionary<Vector2i, BlockedSpaceTile>, float) TransformGridEdges(MapId targetMap, GridId? referenceGrid, List<GridId> localGrids)
+    {
+        Dictionary<Vector2i, BlockedSpaceTile> transformedEdges = new();
+
+        var targetMatrix = Matrix3.Identity;
+        Angle targetAngle = new();
+        var tileSize = (float) DefaultTileSize;
+
+        // if the explosion is centered on some grid (and not just space), get the transforms.
+        if (referenceGrid != null)
+        {
+            var targetGrid = _mapManager.GetGrid(referenceGrid.Value);
+            var xform = Transform(targetGrid.GridEntityId);
+            targetAngle = xform.WorldRotation;
+            targetMatrix = xform.InvWorldMatrix;
+            tileSize = (float) targetGrid.TileSize;
+        }
+
+        var offsetMatrix = Matrix3.Identity;
+        offsetMatrix.R0C2 = tileSize / 2;
+        offsetMatrix.R1C2 = tileSize / 2;
+
+        // Here we can end up with a triple nested for loop:
+        // foreach other grid
+        //   foreach edge tile in that grid
+        //     foreach tile in our grid that touches that tile (vast majority of the time: 1 tile, but could be up to 4)
+
+        HashSet<Vector2i> transformedTiles = new();
+        foreach (var gridToTransform in localGrids)
+        {
+            // we treat the target grid separately
+            if (gridToTransform == referenceGrid)
+                continue;
+
+            if (!_gridEdges.TryGetValue(gridToTransform, out var edges))
+                continue;
+
+            if (!_mapManager.TryGetGrid(gridToTransform, out var grid) ||
+                grid.ParentMapId != targetMap)
+                continue;
+
+            if (grid.TileSize != tileSize)
+            {
+                Logger.Error($"Explosions do not support grids with different grid sizes. GridIds: {gridToTransform} and {referenceGrid}");
+                continue;
+            }
+
+            var xform = EntityManager.GetComponent<TransformComponent>(grid.GridEntityId);
+            var matrix = offsetMatrix * xform.WorldMatrix * targetMatrix;
+            var angle = xform.WorldRotation - targetAngle;
+
+            var (x, y) = angle.RotateVec((tileSize / 4, tileSize / 4));
+
+            foreach (var (tile, dir) in edges)
+            {
+                var center = matrix.Transform(tile);
```

Comment:
Yeah, transforming the grid edges to figure out what parts of space an explosion can/cant go to & when it should transition from space to grids is one of the expensive parts of the explosion tile/area calculation. It only has to do this when the explosion reaches space, and only for grids in-range of the explosion. But for small explosions in space ( & near a grid) this is probably the bottle-neck, as it currently doesn't scale with explosion size (until it reaches other grids/stations). For large explosions the, actuall tile-fill / neighbour finding starts being slower, which in turn is much slower than the eventual processing (destructible deletion & spawning)

I've now updated this with an easy optimisation. Instead of processing each tile, it first checks if the distance of the tile from the epicentre is larger than some distance, and it is it just ignores that tile. So this means that small-medium space-based explosions near grids don't have to deal with all of the grid edges, only those that would ever be in range. The max distance is the same as the one used in the grid lookup to determine what grids could be relevant to an explosion.

This will still iterate over all the edges, but should be much faster than before because for small explosions most of the edges it will just be a very quick Vector2i distance checks. I guess it could still be sped up by chunking the edge data, and then iterating over chunks with a distance-to-chunk-centre check. 

I tried chunking earlier, and it didn't seem to help much with performance, though maybe with a chunk-distance check it might actually be better.